### PR TITLE
Fix null pointer when sending expired token on AST scanner

### PR DESCRIPTION
**Pull request by olgakil**

### CODE REVIEW

**Comment by ghannamz** [comment]
Scan submitted to Checkmarx

**Comment by ghannamz** [comment]
### Checkmarx scan completed
[Full Scan Details](http://EC2AMAZ-JP2I9QF/CxWebClient/ViewerMain.aspx?scanid=1010679&projectid=226)
#### Checkmarx Scan Summary
Severity|Count
---|---
High|0
Medium|11
Low|146
Informational|2

#### Violation Summary
Severity|Count
---|---
Medium|11

#### Details
|Lines|Severity|Category|File|Link|
---|---|---|---|---
[14](https://github.com/checkmarx-ltd/Cx-Client-Common/blob/pr-orly-fix-null-pointer-expired-token/src/main/java/com/cx/restclient/dto/LoginSettings.java#L14) [22](https://github.com/checkmarx-ltd/Cx-Client-Common/blob/pr-orly-fix-null-pointer-expired-token/src/main/java/com/cx/restclient/dto/LoginSettings.java#L22) |Medium|Heap_Inspection|src/main/java/com/cx/restclient/dto/LoginSettings.java|[Checkmarx](http://EC2AMAZ-JP2I9QF/CxWebClient/ViewerMain.aspx?scanid=1010679&projectid=226&pathid=60)
[31](https://github.com/checkmarx-ltd/Cx-Client-Common/blob/pr-orly-fix-null-pointer-expired-token/src/main/java/com/cx/restclient/configuration/CxScanConfig.java#L31) |Medium|Heap_Inspection|src/main/java/com/cx/restclient/configuration/CxScanConfig.java|[Checkmarx](http://EC2AMAZ-JP2I9QF/CxWebClient/ViewerMain.aspx?scanid=1010679&projectid=226&pathid=58)
[13](https://github.com/checkmarx-ltd/Cx-Client-Common/blob/pr-orly-fix-null-pointer-expired-token/src/main/java/com/cx/restclient/dto/LoginRequest.java#L13) |Medium|Heap_Inspection|src/main/java/com/cx/restclient/dto/LoginRequest.java|[Checkmarx](http://EC2AMAZ-JP2I9QF/CxWebClient/ViewerMain.aspx?scanid=1010679&projectid=226&pathid=59)
[13](https://github.com/checkmarx-ltd/Cx-Client-Common/blob/pr-orly-fix-null-pointer-expired-token/src/main/java/com/cx/restclient/osa/dto/LoginRequest.java#L13) |Medium|Heap_Inspection|src/main/java/com/cx/restclient/osa/dto/LoginRequest.java|[Checkmarx](http://EC2AMAZ-JP2I9QF/CxWebClient/ViewerMain.aspx?scanid=1010679&projectid=226&pathid=63)
[120](https://github.com/checkmarx-ltd/Cx-Client-Common/blob/pr-orly-fix-null-pointer-expired-token/src/main/java/com/cx/restclient/ast/AstClient.java#L120) |Medium|Heap_Inspection|src/main/java/com/cx/restclient/ast/AstClient.java|[Checkmarx](http://EC2AMAZ-JP2I9QF/CxWebClient/ViewerMain.aspx?scanid=1010679&projectid=226&pathid=56)
[21](https://github.com/checkmarx-ltd/Cx-Client-Common/blob/pr-orly-fix-null-pointer-expired-token/src/main/java/com/cx/restclient/osa/dto/ScanConfiguration.java#L21) |Medium|Heap_Inspection|src/main/java/com/cx/restclient/osa/dto/ScanConfiguration.java|[Checkmarx](http://EC2AMAZ-JP2I9QF/CxWebClient/ViewerMain.aspx?scanid=1010679&projectid=226&pathid=64)
[10](https://github.com/checkmarx-ltd/Cx-Client-Common/blob/pr-orly-fix-null-pointer-expired-token/src/main/java/com/cx/restclient/dto/ProxyConfig.java#L10) |Medium|Heap_Inspection|src/main/java/com/cx/restclient/dto/ProxyConfig.java|[Checkmarx](http://EC2AMAZ-JP2I9QF/CxWebClient/ViewerMain.aspx?scanid=1010679&projectid=226&pathid=61)
[120](https://github.com/checkmarx-ltd/Cx-Client-Common/blob/pr-orly-fix-null-pointer-expired-token/src/main/java/com/cx/restclient/ast/AstScaClient.java#L120) |Medium|Heap_Inspection|src/main/java/com/cx/restclient/ast/AstScaClient.java|[Checkmarx](http://EC2AMAZ-JP2I9QF/CxWebClient/ViewerMain.aspx?scanid=1010679&projectid=226&pathid=57)
[16](https://github.com/checkmarx-ltd/Cx-Client-Common/blob/pr-orly-fix-null-pointer-expired-token/src/main/java/com/cx/restclient/dto/RemoteSourceRequest.java#L16) |Medium|Heap_Inspection|src/main/java/com/cx/restclient/dto/RemoteSourceRequest.java|[Checkmarx](http://EC2AMAZ-JP2I9QF/CxWebClient/ViewerMain.aspx?scanid=1010679&projectid=226&pathid=62)
[233](https://github.com/checkmarx-ltd/Cx-Client-Common/blob/pr-orly-fix-null-pointer-expired-token/src/main/java/com/cx/restclient/configuration/CxScanConfig.java#L233) |Medium|Privacy_Violation|src/main/java/com/cx/restclient/configuration/CxScanConfig.java|[Checkmarx](http://EC2AMAZ-JP2I9QF/CxWebClient/ViewerMain.aspx?scanid=1010679&projectid=226&pathid=159)

File: src/main/java/com/cx/restclient/httpClient/CxHttpClient.java [hunk 1]
```diff
@@ -493,8 +493,10 @@ public void addCustomHeader(String name, String value) {
         } catch (CxTokenExpiredException ex) {
             if (retry) {
                 logTokenError(httpMethod, statusCode, ex);
-                login(lastLoginSettings);
-                return request(httpMethod, contentType, entity, responseType, expectStatus, failedMsg, isCollection, false);
+                if(lastLoginSettings != null) {
+                    login(lastLoginSettings);
+                    return request(httpMethod, contentType, entity, responseType, expectStatus, failedMsg, isCollection, false);
```
Comment (NimrodGolan) [inline src/main/java/com/cx/restclient/httpClient/CxHttpClient.java:498]: What is the treatment in case lastLoginSettings == null?

Comment (alex-ko-dev) [reply src/main/java/com/cx/restclient/httpClient/CxHttpClient.java:498]: Then we don't retry and just throw an exception.

**Comment by ghannamz** [comment]
Scan submitted to Checkmarx

**Comment by ghannamz** [comment]
### Checkmarx scan completed
[Full Scan Details](http://EC2AMAZ-JP2I9QF/CxWebClient/ViewerMain.aspx?scanid=1010682&projectid=226)
#### Checkmarx Scan Summary
Severity|Count
---|---
High|0
Medium|11
Low|146
Informational|2

#### Violation Summary
Severity|Count
---|---
Medium|11

#### Details
|Lines|Severity|Category|File|Link|
---|---|---|---|---
[14](https://github.com/checkmarx-ltd/Cx-Client-Common/blob/pr-orly-fix-null-pointer-expired-token/src/main/java/com/cx/restclient/dto/LoginSettings.java#L14) [22](https://github.com/checkmarx-ltd/Cx-Client-Common/blob/pr-orly-fix-null-pointer-expired-token/src/main/java/com/cx/restclient/dto/LoginSettings.java#L22) |Medium|Heap_Inspection|src/main/java/com/cx/restclient/dto/LoginSettings.java|[Checkmarx](http://EC2AMAZ-JP2I9QF/CxWebClient/ViewerMain.aspx?scanid=1010682&projectid=226&pathid=95)
[31](https://github.com/checkmarx-ltd/Cx-Client-Common/blob/pr-orly-fix-null-pointer-expired-token/src/main/java/com/cx/restclient/configuration/CxScanConfig.java#L31) |Medium|Heap_Inspection|src/main/java/com/cx/restclient/configuration/CxScanConfig.java|[Checkmarx](http://EC2AMAZ-JP2I9QF/CxWebClient/ViewerMain.aspx?scanid=1010682&projectid=226&pathid=93)
[13](https://github.com/checkmarx-ltd/Cx-Client-Common/blob/pr-orly-fix-null-pointer-expired-token/src/main/java/com/cx/restclient/dto/LoginRequest.java#L13) |Medium|Heap_Inspection|src/main/java/com/cx/restclient/dto/LoginRequest.java|[Checkmarx](http://EC2AMAZ-JP2I9QF/CxWebClient/ViewerMain.aspx?scanid=1010682&projectid=226&pathid=94)
[13](https://github.com/checkmarx-ltd/Cx-Client-Common/blob/pr-orly-fix-null-pointer-expired-token/src/main/java/com/cx/restclient/osa/dto/LoginRequest.java#L13) |Medium|Heap_Inspection|src/main/java/com/cx/restclient/osa/dto/LoginRequest.java|[Checkmarx](http://EC2AMAZ-JP2I9QF/CxWebClient/ViewerMain.aspx?scanid=1010682&projectid=226&pathid=98)
[120](https://github.com/checkmarx-ltd/Cx-Client-Common/blob/pr-orly-fix-null-pointer-expired-token/src/main/java/com/cx/restclient/ast/AstClient.java#L120) |Medium|Heap_Inspection|src/main/java/com/cx/restclient/ast/AstClient.java|[Checkmarx](http://EC2AMAZ-JP2I9QF/CxWebClient/ViewerMain.aspx?scanid=1010682&projectid=226&pathid=91)
[21](https://github.com/checkmarx-ltd/Cx-Client-Common/blob/pr-orly-fix-null-pointer-expired-token/src/main/java/com/cx/restclient/osa/dto/ScanConfiguration.java#L21) |Medium|Heap_Inspection|src/main/java/com/cx/restclient/osa/dto/ScanConfiguration.java|[Checkmarx](http://EC2AMAZ-JP2I9QF/CxWebClient/ViewerMain.aspx?scanid=1010682&projectid=226&pathid=99)
[10](https://github.com/checkmarx-ltd/Cx-Client-Common/blob/pr-orly-fix-null-pointer-expired-token/src/main/java/com/cx/restclient/dto/ProxyConfig.java#L10) |Medium|Heap_Inspection|src/main/java/com/cx/restclient/dto/ProxyConfig.java|[Checkmarx](http://EC2AMAZ-JP2I9QF/CxWebClient/ViewerMain.aspx?scanid=1010682&projectid=226&pathid=96)
[120](https://github.com/checkmarx-ltd/Cx-Client-Common/blob/pr-orly-fix-null-pointer-expired-token/src/main/java/com/cx/restclient/ast/AstScaClient.java#L120) |Medium|Heap_Inspection|src/main/java/com/cx/restclient/ast/AstScaClient.java|[Checkmarx](http://EC2AMAZ-JP2I9QF/CxWebClient/ViewerMain.aspx?scanid=1010682&projectid=226&pathid=92)
[16](https://github.com/checkmarx-ltd/Cx-Client-Common/blob/pr-orly-fix-null-pointer-expired-token/src/main/java/com/cx/restclient/dto/RemoteSourceRequest.java#L16) |Medium|Heap_Inspection|src/main/java/com/cx/restclient/dto/RemoteSourceRequest.java|[Checkmarx](http://EC2AMAZ-JP2I9QF/CxWebClient/ViewerMain.aspx?scanid=1010682&projectid=226&pathid=97)
[233](https://github.com/checkmarx-ltd/Cx-Client-Common/blob/pr-orly-fix-null-pointer-expired-token/src/main/java/com/cx/restclient/configuration/CxScanConfig.java#L233) |Medium|Privacy_Violation|src/main/java/com/cx/restclient/configuration/CxScanConfig.java|[Checkmarx](http://EC2AMAZ-JP2I9QF/CxWebClient/ViewerMain.aspx?scanid=1010682&projectid=226&pathid=75)

**Comment by ghannamz** [comment]
Scan submitted to Checkmarx

**Comment by ghannamz** [comment]
### Checkmarx scan completed
[Full Scan Details](http://EC2AMAZ-JP2I9QF/CxWebClient/ViewerMain.aspx?scanid=1010700&projectid=226)
#### Checkmarx Scan Summary
Severity|Count
---|---
High|0
Medium|11
Low|146
Informational|2

#### Violation Summary
Severity|Count
---|---
Medium|11

#### Details
|Lines|Severity|Category|File|Link|
---|---|---|---|---
[14](https://github.com/checkmarx-ltd/Cx-Client-Common/blob/pr-orly-fix-null-pointer-expired-token/src/main/java/com/cx/restclient/dto/LoginSettings.java#L14) [22](https://github.com/checkmarx-ltd/Cx-Client-Common/blob/pr-orly-fix-null-pointer-expired-token/src/main/java/com/cx/restclient/dto/LoginSettings.java#L22) |Medium|Heap_Inspection|src/main/java/com/cx/restclient/dto/LoginSettings.java|[Checkmarx](http://EC2AMAZ-JP2I9QF/CxWebClient/ViewerMain.aspx?scanid=1010700&projectid=226&pathid=95)
[31](https://github.com/checkmarx-ltd/Cx-Client-Common/blob/pr-orly-fix-null-pointer-expired-token/src/main/java/com/cx/restclient/configuration/CxScanConfig.java#L31) |Medium|Heap_Inspection|src/main/java/com/cx/restclient/configuration/CxScanConfig.java|[Checkmarx](http://EC2AMAZ-JP2I9QF/CxWebClient/ViewerMain.aspx?scanid=1010700&projectid=226&pathid=93)
[13](https://github.com/checkmarx-ltd/Cx-Client-Common/blob/pr-orly-fix-null-pointer-expired-token/src/main/java/com/cx/restclient/dto/LoginRequest.java#L13) |Medium|Heap_Inspection|src/main/java/com/cx/restclient/dto/LoginRequest.java|[Checkmarx](http://EC2AMAZ-JP2I9QF/CxWebClient/ViewerMain.aspx?scanid=1010700&projectid=226&pathid=94)
[13](https://github.com/checkmarx-ltd/Cx-Client-Common/blob/pr-orly-fix-null-pointer-expired-token/src/main/java/com/cx/restclient/osa/dto/LoginRequest.java#L13) |Medium|Heap_Inspection|src/main/java/com/cx/restclient/osa/dto/LoginRequest.java|[Checkmarx](http://EC2AMAZ-JP2I9QF/CxWebClient/ViewerMain.aspx?scanid=1010700&projectid=226&pathid=98)
[120](https://github.com/checkmarx-ltd/Cx-Client-Common/blob/pr-orly-fix-null-pointer-expired-token/src/main/java/com/cx/restclient/ast/AstClient.java#L120) |Medium|Heap_Inspection|src/main/java/com/cx/restclient/ast/AstClient.java|[Checkmarx](http://EC2AMAZ-JP2I9QF/CxWebClient/ViewerMain.aspx?scanid=1010700&projectid=226&pathid=91)
[21](https://github.com/checkmarx-ltd/Cx-Client-Common/blob/pr-orly-fix-null-pointer-expired-token/src/main/java/com/cx/restclient/osa/dto/ScanConfiguration.java#L21) |Medium|Heap_Inspection|src/main/java/com/cx/restclient/osa/dto/ScanConfiguration.java|[Checkmarx](http://EC2AMAZ-JP2I9QF/CxWebClient/ViewerMain.aspx?scanid=1010700&projectid=226&pathid=99)
[10](https://github.com/checkmarx-ltd/Cx-Client-Common/blob/pr-orly-fix-null-pointer-expired-token/src/main/java/com/cx/restclient/dto/ProxyConfig.java#L10) |Medium|Heap_Inspection|src/main/java/com/cx/restclient/dto/ProxyConfig.java|[Checkmarx](http://EC2AMAZ-JP2I9QF/CxWebClient/ViewerMain.aspx?scanid=1010700&projectid=226&pathid=96)
[120](https://github.com/checkmarx-ltd/Cx-Client-Common/blob/pr-orly-fix-null-pointer-expired-token/src/main/java/com/cx/restclient/ast/AstScaClient.java#L120) |Medium|Heap_Inspection|src/main/java/com/cx/restclient/ast/AstScaClient.java|[Checkmarx](http://EC2AMAZ-JP2I9QF/CxWebClient/ViewerMain.aspx?scanid=1010700&projectid=226&pathid=92)
[16](https://github.com/checkmarx-ltd/Cx-Client-Common/blob/pr-orly-fix-null-pointer-expired-token/src/main/java/com/cx/restclient/dto/RemoteSourceRequest.java#L16) |Medium|Heap_Inspection|src/main/java/com/cx/restclient/dto/RemoteSourceRequest.java|[Checkmarx](http://EC2AMAZ-JP2I9QF/CxWebClient/ViewerMain.aspx?scanid=1010700&projectid=226&pathid=97)
[233](https://github.com/checkmarx-ltd/Cx-Client-Common/blob/pr-orly-fix-null-pointer-expired-token/src/main/java/com/cx/restclient/configuration/CxScanConfig.java#L233) |Medium|Privacy_Violation|src/main/java/com/cx/restclient/configuration/CxScanConfig.java|[Checkmarx](http://EC2AMAZ-JP2I9QF/CxWebClient/ViewerMain.aspx?scanid=1010700&projectid=226&pathid=75)

**Pull request merged**